### PR TITLE
portable: introduce SYSEXT_ fields to identify sysexts, and include more metadata in log messages via LogExtraFields=

### PR DESCRIPTION
**Pull request by bluca**
Document that `SYSEXT_` + os-release fields can be used to identify the extensions, and append this and other metadata to the log messages

### CODE REVIEW

File: src/portable/portable.c [hunk 1]
```diff
@@ -1005,18 +1005,26 @@ static int install_chroot_dropin(
                                "LogExtraFields=PORTABLE=", base_name, "\n"))
                         return -ENOMEM;
 
+                if (!ordered_hashmap_isempty(extension_images))
+                        if (!strextend(&text, "LogExtraFields=PORTABLE_ROOT=", image_path, "\n"))
```
Comment (poettering) [inline src/portable/portable.c]: why did you name this "_ROOT="? Why not PORTABLE_PATH= or so?

Also, shouldn't this be: `m->image_path ?: image_path`?

File: src/portable/portable.c [hunk 2]
```diff
@@ -1005,18 +1005,26 @@ static int install_chroot_dropin(
...
                                                    FLAGS_SET(flags, PORTABLE_FORCE_SYSEXT) ?
-                                                       ":x-systemd.relax-extension-release-check" :
-                                                       "",
-                                               "\n"))
+                                                       ":x-systemd.relax-extension-release-check\n" :
+                                                       "\n",
+                                               /* In PORTABLE= we list the 'main' image name for this unit
+                                                * (the image where the unit was extracted from), but we are
+                                                * stacking multiple images, so list those too. */
+                                               "LogExtraFields=PORTABLE_EXTENSION=", ext->path, "\n"))
```
Comment (poettering) [inline src/portable/portable.c]: `PORTABLE_EXTENSION_PATH=`?

File: src/portable/portable.c [hunk 3]
```diff
@@ -951,11 +951,98 @@ static int make_marker_text(const char *image_path, OrderedHashmap *extension_im
...
+                /* The extension-release filename should always start with extension-release. */
+                assert(image_name);
+
+                prefix = strjoina(image_name, ":");
+        } else
+                prefix = "";
+
+        r = parse_env_file_fd(release->fd,
+                              release->name,
+                              sysext_prefix ? "SYSEXT_ID" : "ID",
```
Comment (poettering) [inline src/portable/portable.c]: use:

```c
static const char *const*const field_names[2] = {
        [false] = { "ID", "VERSION_ID", "IMAGE_VERSION", "IMAGE_ID", "BUILD_ID", NULL },
        [true]  = { "SYSEXT_ID", …, NULL },
};
```

and then:

```
r = load_env_file_pairs_fd(…, &l);

STRV_FOREACH_PAIR(k, v, l) {
        if (!strv_find(field_names[sysext_prefix], k)) 
                continue;
        
        if (!strextend(…)…
}
```

you get the idea.

File: src/portable/portable.c [hunk 4]
```diff
@@ -951,11 +951,98 @@ static int make_marker_text(const char *image_path, OrderedHashmap *extension_im
         return 0;
 }
 
+/* 'output=' is always appended to, even if there are ENOMEM errors */
+static int append_release_info_to_unit(char **output, const PortableMetadata *release, bool sysext_prefix) {
```
Comment (poettering) [inline src/portable/portable.c]: hmm, maybe use ImageClass for the 3rd arg, instead of "bool". I figure sooner or later we want this for the syscfg stuff anyway, too. And using booleans for choosing between two sets of names is a bit icky anyway

File: src/portable/portable.c (hunk 1 above)
Comment (bluca) [reply src/portable/portable.c]: Because it's what's used for RootImage[/Directory], it's intentional to encode that information. Just "PATH" is ambiguous, as it doesn't tell you for what it was used.

File: src/portable/portable.c (hunk 2 above)
Comment (bluca) [reply src/portable/portable.c]: as above

File: src/portable/portable.c (hunk 1 above)
Comment (poettering) [reply src/portable/portable.c]: i cannot parse that? how is that used for RootImage=?

I wuld assume ROOT= would actually refer to a dir, not an image, hence i am irritated by this.

what's the usecase for this btw? why isn't the basename of the image enough, i.e. the current PORTABLE= field? why do you need the path too?

File: src/portable/portable.c (hunk 2 above)
Comment (poettering) [reply src/portable/portable.c]: i don't follow? this is a a full path, but claling it path is ambiguous? and just calling it "extension" is better?

File: src/portable/portable.c (hunk 1 above)
Comment (bluca) [reply src/portable/portable.c]: Maybe an example would help to clarify - this is attached with 2 extensions and a base, and the unit file is extracted from the app0 extension, 

```
# Drop-in created for image '/home/bluca/git/systemd/base.raw:/home/bluca/git/systemd/app0.raw:/home/bluca/git/systemd/app1.raw', do not edit.

[Service]
RootImage=/home/bluca/git/systemd/base.raw
Environment=PORTABLE=app0.raw
BindReadOnlyPaths=/etc/os-release:/run/host/os-release
LogExtraFields=PORTABLE=app0.raw
LogExtraFields=VERSION_ID=10
LogExtraFields=ID=debian
LogExtraFields=PORTABLE_ROOT=/home/bluca/git/systemd/base.raw

ExtensionImages=/home/bluca/git/systemd/app0.raw
LogExtraFields=PORTABLE_EXTENSION=/home/bluca/git/systemd/app0.raw
LogExtraFields=SYSEXT_VERSION_ID=app0:0

ExtensionImages=/home/bluca/git/systemd/app1.raw
LogExtraFields=PORTABLE_EXTENSION=/home/bluca/git/systemd/app1.raw
LogExtraFields=SYSEXT_VERSION_ID=app1:1
```

- PORTABLE= is the image _name_ of the source of the unit file, so in this case it's `app0.raw` as that's where the unit is from
- PORTABLE_ROOT points to the base image, which is used as RootImage= or RootDirectory= - the distinction is not important for this purpose, the important thing is what is used as a base, the name itself encodes whether it's a directory or an image (e.g. `base/` vs `base.raw`)
- PORTABLE_EXTENSION points to each extension image/directory, again used as ExtensionImages= or ExtensionDirectories= same as above

The use case is that it's useful to me to know when looking retrospectively at old, archived log messages for the unit, exactly on which image they are running on.

Having the full path is not super important, although it might be useful in some cases, and it's more work to extract the base name than it is to just use it as-is, but if you insist I can rework that, it's not a biggie

File: src/portable/portable.c (hunk 2 above)
Comment (bluca) [reply src/portable/portable.c]: it's not about the path but about the root vs extension, full example above hopefully clarifies a bit

Comment (bluca) [reply src/portable/portable.c]: ie: I'm fine with either adding _PATH or switching to just the base name

Comment (bluca) [reply src/portable/portable.c]: removed the full path

File: src/portable/portable.c (hunk 1 above)
Comment (bluca) [reply src/portable/portable.c]: in the interest of saving time I've switched to using the basename everywhere rather than the full path, does this work better now? Example:

```
[Service]
RootImage=/home/bluca/git/systemd/base.raw
Environment=PORTABLE=app0.raw
BindReadOnlyPaths=/etc/os-release:/run/host/os-release
LogExtraFields=PORTABLE=app0.raw
LogExtraFields=PORTABLE_ROOT=base.raw
LogExtraFields=VERSION_ID=10
LogExtraFields=ID=debian

ExtensionImages=/home/bluca/git/systemd/app0.raw
LogExtraFields=PORTABLE_EXTENSION=app0.raw
LogExtraFields=SYSEXT_VERSION_ID=app0:0

ExtensionImages=/home/bluca/git/systemd/app1.raw
LogExtraFields=PORTABLE_EXTENSION=app1.raw
LogExtraFields=SYSEXT_VERSION_ID=app1:1
```

File: src/basic/env-file.c [hunk 5]
```diff
@@ -494,6 +495,25 @@ int load_env_file_pairs(FILE *f, const char *fname, char ***ret) {
...
+
+        fd_ro = fd_reopen(fd, O_CLOEXEC | O_RDONLY);
+        if (fd_ro < 0)
+                return fd_ro;
+
+        f = fdopen(fd_ro, "re");
+        if (!f)
+                return -errno;
+
+        TAKE_FD(fd_ro);
```
Comment (poettering) [inline src/basic/env-file.c]: please rebase on #26952 and use the new helper here.

File: src/test/test-fileio.c [hunk 6]
```diff
@@ -492,6 +492,23 @@ TEST(load_env_file_pairs) {
...
         assert_se(r == 0);
 
+        r = load_env_file_pairs_fd(fd, fn, &l);
+        assert_se(r >= 0);
+
+        assert_se(strv_length(l) == 14);
+        STRV_FOREACH_PAIR(k, v, l) {
+                assert_se(STR_IN_SET(*k, "NAME", "ID", "PRETTY_NAME", "ANSI_COLOR", "HOME_URL", "SUPPORT_URL", "BUG_REPORT_URL"));
+                printf("%s=%s\n", *k, *v);
+                if (streq(*k, "NAME")) assert_se(streq(*v, "Arch Linux"));
```
Comment (poettering) [inline src/test/test-fileio.c]: this is weird.

just write:

```c
assert_se(!streq(*k, "NAME") || streq(*v, "Arch linux"));

File: src/test/test-fileio.c [hunk 7]
```diff
@@ -492,6 +492,23 @@ TEST(load_env_file_pairs) {
...
+                if (streq(*k, "ID")) assert_se(streq(*v, "arch"));
+                if (streq(*k, "PRETTY_NAME")) assert_se(streq(*v, "Arch Linux"));
+                if (streq(*k, "ANSI_COLOR")) assert_se(streq(*v, "0;36"));
+                if (streq(*k, "HOME_URL")) assert_se(streq(*v, "https://www.archlinux.org/"));
+                if (streq(*k, "SUPPORT_URL")) assert_se(streq(*v, "https://bbs.archlinux.org/"));
+                if (streq(*k, "BUG_REPORT_URL")) assert_se(streq(*v, "https://bugs.archlinux.org/"));
```
Comment (poettering) [inline src/test/test-fileio.c]: oh, this is a copy of the code below? please add a helper call for that, and share the verification. no copy/paste of this mess please

File: src/basic/env-file.c [hunk 8]
```diff
@@ -494,6 +495,25 @@ int load_env_file_pairs(FILE *f, const char *fname, char ***ret) {
         return 0;
 }
 
+int load_env_file_pairs_fd(int fd, const char *fname, char ***ret) {
```
Comment (poettering) [inline src/basic/env-file.c:491]: there's one case where load_env_pairs() is currently used in extract_image_and_extensions() which should be ported over to this as well

File: src/portable/portable.c [hunk 9]
```diff
@@ -951,11 +951,68 @@ static int make_marker_text(const char *image_path, OrderedHashmap *extension_im
         return 0;
 }
 
+/* 'output=' is always appended to, even if there are ENOMEM errors */
+static int append_release_info_to_unit(char **output, const PortableMetadata *release, ImageClass type) {
+        static const char *const field_names[_IMAGE_CLASS_MAX][6]= {
+                 {}, /* IMAGE_MACHINE is not expected here, the assert below enforces it */
+                 { "ID", "VERSION_ID", "IMAGE_VERSION", "IMAGE_ID", "BUILD_ID", NULL },
+                 { "SYSEXT_ID", "SYSEXT_VERSION_ID", "SYSEXT_IMAGE_VERSION", "SYSEXT_IMAGE_ID", "SYSEXT_BUILD_ID", NULL },
```
Comment (poettering) [inline src/portable/portable.c]: `[IMAGE_EXTENSION] = { … }`

File: src/portable/portable.c [hunk 10]
```diff
@@ -951,11 +951,68 @@ static int make_marker_text(const char *image_path, OrderedHashmap *extension_im
         return 0;
 }
 
+/* 'output=' is always appended to, even if there are ENOMEM errors */
+static int append_release_info_to_unit(char **output, const PortableMetadata *release, ImageClass type) {
+        static const char *const field_names[_IMAGE_CLASS_MAX][6]= {
+                 {}, /* IMAGE_MACHINE is not expected here, the assert below enforces it */
+                 { "ID", "VERSION_ID", "IMAGE_VERSION", "IMAGE_ID", "BUILD_ID", NULL },
```
Comment (poettering) [inline src/portable/portable.c]: [IMAGE_PORTABLE] = { … },

File: src/portable/portable.c [hunk 11]
```diff
@@ -951,11 +951,68 @@ static int make_marker_text(const char *image_path, OrderedHashmap *extension_im
         return 0;
 }
 
+/* 'output=' is always appended to, even if there are ENOMEM errors */
+static int append_release_info_to_unit(char **output, const PortableMetadata *release, ImageClass type) {
+        static const char *const field_names[_IMAGE_CLASS_MAX][6]= {
+                 {}, /* IMAGE_MACHINE is not expected here, the assert below enforces it */
```
Comment (poettering) [inline src/portable/portable.c]: just drop this line, will NULL initialized anyway.

instead maybe add `assert(!strv_isempty(field_names[type]));` below?

File: src/portable/portable.c [hunk 12]
```diff
@@ -1039,6 +1100,19 @@ static int install_chroot_dropin(
...
+
+                                /* Look for image/version identifiers in the extension release files. We
+                                 * look for all possible IDs, but typically only 1 or 2 will be set, so
+                                 * the number of fields added shouldn't be too large. We prefix the DDI
+                                 * name to the value, so that we can add the same field multiple times and
+                                 * still be able to identify what applies to what. */
+                                PortableMetadata *extension_release = ordered_hashmap_get(extension_releases,
+                                                                                          ext->name);
+                                r = append_release_info_to_unit(&text,
+                                                                extension_release,
```
Comment (poettering) [inline src/portable/portable.c]: i think this call would be more readable like this:

```c
r = append_release_info_to_unit(
                &text,
                ordered_hashmap_get(extension_releases, ext->name),
                IMAGE_EXTENSION);
```

File: src/portable/portable.c [hunk 13]
```diff
@@ -951,11 +951,68 @@ static int make_marker_text(const char *image_path, OrderedHashmap *extension_im
...
 
+/* 'output=' is always appended to, even if there are ENOMEM errors */
+static int append_release_info_to_unit(char **output, const PortableMetadata *release, ImageClass type) {
+        static const char *const field_names[_IMAGE_CLASS_MAX][6]= {
+                 {}, /* IMAGE_MACHINE is not expected here, the assert below enforces it */
+                 { "ID", "VERSION_ID", "IMAGE_VERSION", "IMAGE_ID", "BUILD_ID", NULL },
+                 { "SYSEXT_ID", "SYSEXT_VERSION_ID", "SYSEXT_IMAGE_VERSION", "SYSEXT_IMAGE_ID", "SYSEXT_BUILD_ID", NULL },
+        };
+        _cleanup_strv_free_ char **fields = NULL;
+        _cleanup_free_ char *basename = NULL;
```
Comment (poettering) [inline src/portable/portable.c]: move to innermost scope?

File: src/test/test-fileio.c (hunk 7 above)
Comment (bluca) [reply src/test/test-fileio.c]: I didn't copy/paste it, copilot did!

File: docs/PORTABLE_SERVICES.md [hunk 14]
```diff
@@ -345,15 +345,28 @@ was loaded from. In case extensions are used, additionally there will be a
...
+extension-releases), and a base layer `base.raw` (with `VERSION_ID=10` and
+`ID=debian` in os-release), will create log entries with the following fields:
 
 ```
 PORTABLE=app0.raw
 PORTABLE_ROOT=base.raw
 PORTABLE_EXTENSION=app0.raw
 PORTABLE_EXTENSION=app1.raw
+VERSION_ID=10
+ID=debian
```
Comment (poettering) [inline docs/PORTABLE_SERVICES.md]: hmm, so we end up adding this into all log messages written by services. we probably should be a bit careful with namespace conflicts there. i.e. not use super duper generic field names such as "ID=" for this, because that sounds like something apps might want to use for themselves. This is in particular relevant since this ID field here is an OS ID, but that is not clear at all if you just look at the field name.

hence, maybe prefix these fields with OS_RELEASE_ or so? That would make them a lot more specific and clashes much less likely. Moreover, it would leave the door open that we eventually attach these fields entirely generically and automatically also for non-portable services.

**Comment by poettering** [comment]
looks great. just concerned about the super generic journal field name ID=, which we really should avoid. otherwise is good to go

File: docs/PORTABLE_SERVICES.md (hunk 14 above)
Comment (bluca) [reply docs/PORTABLE_SERVICES.md]: yes I was thinking the same, and also the format is a bit unwieldly, and we get a lot of fields added, so instead I'll add a single field per image, called PORTABLE_ID and compose the value from an ID and a VERSION_ID, depending on what's available

**Comment by bluca** [comment]
> looks great. just concerned about the super generic journal field name ID=, which we really should avoid. otherwise is good to go

Changed to assemble a single field per image, `PORTABLE_ID=` (or `PORTABLE_ROOT_ID=` plus `PORTABLE_EXTENSION_ID=` in case of layering), composed from an id and version fields, if found. So we have a single field rather than N, and they are namespaced. Example in the commit message. PTAL.

File: src/portable/portable.c [hunk 15]
```diff
@@ -994,15 +1058,30 @@ static int install_chroot_dropin(
                                "LogExtraFields=PORTABLE=", base_name, "\n"))
                         return -ENOMEM;
 
+                /* If we have a single image then PORTABLE= will point to it, so we add PORTABLE_ID= with
+                 * the os-release fields and we are done. But if we have extensions, PORTABLE= will point
+                 * to the image where the current unit was found in. So we also list PORTABLE_ROOT= and
+                 * PORTABLE_ROOT_ID= for the base image, and PORTABLE_EXTENSION= and PORTABLE_EXTENSION_ID=
+                 * for each extension, so that all needed metadata is available. */
                 if (!ordered_hashmap_isempty(extension_images)) {
```
Comment (poettering) [inline src/portable/portable.c]: invert the if check, i.e. put the empty case first, and the non-empty case second.

File: src/portable/portable.c [hunk 16]
```diff
@@ -994,15 +1058,30 @@ static int install_chroot_dropin(
...
+                                       "LogExtraFields=PORTABLE_ROOT=", root_base_name, "\n"))
                                 return -ENOMEM;
+
+                        r = append_release_log_fields(os_release, IMAGE_PORTABLE, "PORTABLE_ROOT_ID", &text);
+                        if (r < 0)
+                                return r;
+                } else {
+                        r = append_release_log_fields(os_release, IMAGE_PORTABLE, "PORTABLE_ID", &text);
+                        if (r < 0)
+                                return r;
```
Comment (poettering) [inline src/portable/portable.c]: the last two lines of both if branches are the same. move out of the if branches

Comment (bluca) [reply src/portable/portable.c]: updated

File: src/portable/portable.c [hunk 17]
```diff
@@ -940,11 +940,75 @@ static int make_marker_text(const char *image_path, OrderedHashmap *extension_im
...
+                if (id)
+                        break;
+        }
+
+        /* Then the version, same logic, prefer the more specific one */
+        STRV_FOREACH(k, (char *const *)field_versions[type]) {
+                version = strv_env_pairs_get(fields, *k);
+                if (version)
+                        break;
+        }
```
Comment (poettering) [inline src/portable/portable.c]: doesn't matter, but you could make this a helper call: find_first_field() or so that takes two char** arrays, and does this.

File: src/portable/portable.c [hunk 18]
```diff
@@ -994,16 +1058,29 @@ static int install_chroot_dropin(
...
 
-                if (!ordered_hashmap_isempty(extension_images)) {
+                /* If we have a single image then PORTABLE= will point to it, so we add PORTABLE_ID= with
+                 * the os-release fields and we are done. But if we have extensions, PORTABLE= will point
+                 * to the image where the current unit was found in. So we also list PORTABLE_ROOT= and
+                 * PORTABLE_ROOT_ID= for the base image, and PORTABLE_EXTENSION= and PORTABLE_EXTENSION_ID=
+                 * for each extension, so that all needed metadata is available. */
+                if (ordered_hashmap_isempty(extension_images)) {
+                        r = append_release_log_fields(os_release, IMAGE_PORTABLE, "PORTABLE_ID", &text);
+                } else {
```
Comment (poettering) [inline src/portable/portable.c]: no {} around single-line if blocks,as per coding style

Comment (poettering) [reply src/portable/portable.c]: hmm, the combination of a basename and a version should this really be called "ID"? rpm calls this "nevra" (i.e. name, epoch, version, release, architecture). Maybe we should call this something like "object"? or be literal, i.e. "PORTABLE_NAME_AND_VERSION"? PORTABLE_NAME_VERSION, PORTABLE_NVER? PORTABLE_NAMVER? or something like this?

so far in /etc/os-release we used the "_ID" suffix a bit chaotic but usually it's used as a low-level identifier rather than something to expose to the user. it's never a combination of things. This here seems to be quite a different use though, hence we should really pick a different moniker.

PORTABLE_MONIKER? PORTABLE_NAME?

kinda like "moniker" for this...

File: man/os-release.xml [hunk 19]
```diff
@@ -111,6 +111,11 @@
       <varname>VERSION_ID=</varname> exists and matches. This ensures ABI/API compatibility between the
       layers and prevents merging of an incompatible image in an overlay.</para>
 
+      <para>In order to identify the extension image itself, the same fields defined below can be added to the
+      <filename>extension-release</filename> file with a <varname>SYSEXT_</varname> prefix (to disambiguate
+      from fields used to match on the base image). E.g.: <varname>SYSEXT_ID=myext</varname>,
+      <varname>SYSEXT_VERSION_ID=1.2.3</varname>.</para>
```
Comment (poettering) [inline man/os-release.xml:117]: still would prefer if we'd split this out

**Review by poettering — CHANGES_REQUESTED** [review]

File: man/os-release.xml (hunk 19 above)
Comment (bluca) [reply man/os-release.xml:117]: This is just an example of a mapping 1:1 from the existing os-release, it's not the new log field name

File: src/portable/portable.c (hunk 18 above)
Comment (bluca) [reply src/portable/portable.c]: changing _ID -> _NAME_AND_VERSION

File: src/portable/portable.c (hunk 17 above)
Comment (bluca) [reply src/portable/portable.c]: it's really a corner case of a corner case, let's leave the helper out until there's a separate use case also appears somewhere else

Comment (poettering) [reply src/portable/portable.c]: well, you already have the same loop twice here. thats why i am suggesting to make it a helper call...

Comment (bluca) [reply src/portable/portable.c]: added

File: src/portable/portable.c [hunk 20]
```diff
@@ -994,19 +994,41 @@ static int install_chroot_dropin(
...
-                        ORDERED_HASHMAP_FOREACH(ext, extension_images)
+                        ORDERED_HASHMAP_FOREACH(ext, extension_images) {
+                                _cleanup_free_ char *extension_base_name = NULL;
+
+                                r = path_extract_filename(ext->path, &extension_base_name);
+                                if (r < 0)
+                                        return log_debug_errno(r, "Failed to extract basename from '%s': %m", ext->path);
+
                                 if (!strextend(&text,
+                                               "\n",
```
Comment (keszybz) [inline src/portable/portable.c:1087]: Why `\n` here? There's already one added at line 1004, no?

File: docs/PORTABLE_SERVICES.md [hunk 21]
```diff
@@ -335,6 +335,27 @@ service data may be placed on the host file system. Use `StateDirectory=` in
...
+
+For example, a portable service `app0` using two extensions `app0.raw` and
+`app1.raw`, and a base layer `base.raw`, will create log entries with the
+following fields:
+
+```
+PORTABLE=app0.raw
+PORTABLE_ROOT=base.raw
+PORTABLE_EXTENSION=app0.raw
+PORTABLE_EXTENSION=app1.raw
```
Comment (keszybz) [inline docs/PORTABLE_SERVICES.md]: I'm also worried about the non-uniqueness of this name that @poettering raised earlier.
There are two general types of fields: app-specific, where short names are OK because it is only the app that needs to coordinate with itself, and system-wide, which are appended to many or a large fraction of log messages. For those, it is important to have namaspacing so that conflicts don't occur, and that it's clear what the field is for.

"PORTABLE_" works as a kind of a namespace, but it's also a normal word, so the namespacing is "weak". ~~What about changing that to "SYSTEMD_PORTABLE_"? Then the chances of misunderstanding or conflict go down to negligible.~~ I guess it's good enough. Considering that you also want to have envvars with the same contents, it makes sense to use the same names.

File: docs/PORTABLE_SERVICES.md [hunk 22]
```diff
@@ -345,15 +345,32 @@ was loaded from. In case extensions are used, additionally there will be a
 (i.e.: `RootImage=` or `RootDirectory=`), and one `PORTABLE_EXTENSION=` field per
 each extension image used.
 
+The os-release file from the portable image will be parsed and added as structured
```
Comment (keszybz) [inline docs/PORTABLE_SERVICES.md]: Markup around `os-release` missing.

File: docs/PORTABLE_SERVICES.md [hunk 23]
```diff
@@ -345,15 +345,32 @@ was loaded from. In case extensions are used, additionally there will be a
...
+metadata to the journal log entries by for an ID between `IMAGE_ID` and `ID` in this
```
Comment (keszybz) [inline docs/PORTABLE_SERVICES.md]: "by for" ?

I don't grok the sentence.

File: docs/PORTABLE_SERVICES.md [hunk 24]
```diff
@@ -345,15 +345,32 @@ was loaded from. In case extensions are used, additionally there will be a
...
+`SYSEXT_`, are parsed from each extension-release file, and are appended to the
+journal as log entries, using `PORTABLE_EXTENSION_NAME_AND_VERSION=` as the field
+name. The base layer's field will be named `PORTABLE_ROOT_NAME_AND_VERSION=` instead
+of `PORTABLE_NAME_AND_VERSION=` in this case.
+
 For example, a portable service `app0` using two extensions `app0.raw` and
-`app1.raw`, and a base layer `base.raw`, will create log entries with the
-following fields:
+`app1.raw` (with `SYSEXT_ID=app`, and `SYSEXT_VERSION_ID=` `0` and `1` in their
+respective extension-releases), and a base layer `base.raw` (with `VERSION_ID=10` and
```
Comment (keszybz) [inline docs/PORTABLE_SERVICES.md:364]: Markup around `extension-release`?

File: src/portable/portable.c [hunk 25]
```diff
@@ -940,11 +940,67 @@ static int make_marker_text(const char *image_path, OrderedHashmap *extension_im
         return 0;
 }
 
+static int append_release_log_fields(
+                const PortableMetadata *release,
+                ImageClass type,
+                const char *field_name,
+                char **ret_text) {
```
Comment (keszybz) [inline src/portable/portable.c]: Hmm, this should not be `ret_`, because it's an input-output field. 
Maybe also move that param to the first position.

**Comment by poettering** [comment]
hmm, ppc ci:

```
TEST-29-PORTABLE:                   FAIL     (1205 s)
```

**Comment by poettering** [comment]
(the other CI failures appear unrelated)

File: src/portable/portablectl.c [hunk 26]
```diff
@@ -400,38 +400,51 @@ static int inspect_image(int argc, char *argv[], void *userdata) {
...
+                                                   "SYSEXT_IMAGE_VERSION", &image_version,
+                                                   "SYSEXT_PRETTY_NAME", &pretty_os,
                                                    "SYSEXT_SCOPE", &sysext_scope,
                                                    "SYSEXT_LEVEL", &sysext_level,
+                                                   "ID", &id,
+                                                   "VERSION_ID", &version_id,
                                                    "PORTABLE_PRETTY_NAME", &pretty_portable,
-                                                   "PORTABLE_PREFIXES", &portable_prefixes,
-                                                   "PRETTY_NAME", &pretty_os);
+                                                   "PORTABLE_PREFIXES", &portable_prefixes);
```
Comment (keszybz) [inline src/portable/portablectl.c:423]: Hmm, this changes what fields are displayed… I'm not sure if this is intentional; it should be a separate commit. (Or somehow explained in the commit message.)

**Review by keszybz — CHANGES_REQUESTED** [review]
Some more questions. Sorry for the late review.

File: src/portable/portablectl.c (hunk 26 above)
Comment (bluca) [reply src/portable/portablectl.c:423]: It is part of the same change adding the new fields, as before the output was essentially useless - a long list of `(n/a)`. But now we have the new fields that are specific for extensions and we can use them.

File: src/portable/portable.c (hunk 20 above)
Comment (bluca) [reply src/portable/portable.c:1087]: To separate the blocks. This way there's the root block, then each extension is in a separate new-line separated block. It looks much more readable to me.

**Comment by bluca** [comment]
> hmm, ppc ci:
> 
> ```
> TEST-29-PORTABLE:                   FAIL     (1205 s)
> ```

It's unrelated, see: https://github.com/systemd/systemd/issues/25091

File: docs/PORTABLE_SERVICES.md (hunk 21 above)
Comment (bluca) [reply docs/PORTABLE_SERVICES.md]: Yeah we were already using PORTABLE*, so just went on with the same prefix. We have been using "portable" for this feature for a while, and these are added only when portabled is used, so the namespacing should be ok, it is not added to generic services, only those attached by portabled.
